Defensive & Offensive Programming (#166)
memfault/interrupt: _drafts/offensive-programming.md

The firmware _knows_ there is an issue, and it likely knows _what the issue is_,
but yet it doesn't help solve the issue. The software was built with "Defensive
Programming" practices in mind, which allows for the firmware to keep running
even when incorrect behavior occurs.

Usually, I would prefer the firmware crash, print a useful error, and guide me
exactly to where the issue is. This the guiding principle of "Offensive
Programming", which goes one step further than "Defensive Programming".

<!-- excerpt start -->

In this article, we'll dive into what defensive and offensive programming are,
where defensive programming falls short, how developers should think about using
them within embedded systems, and how the techniques of offensive programming
can surface bugs and help developers root cause them immediately at runtime.

<!-- excerpt end -->

By taking defensive and offensive programming to the extreme, you'll be able to
track down those 1 in 1,000 hour bugs, effeciently root cause them, and keep
your end-users happy.

{% include newsletter.html %}

{% include toc.html %}

## Defensive Programming

> Defensive programming (referred to from here on as DefP) is the practice of

Author: Removed abbreviations. Thanks for the push.

want to avoid.

On top of preventing your firmware from running in an unpredictable state,
triggering asserts and faults at the exact moment of the bug helps developers
track down these issues! If your devices are connected to a debugger while an
assert or fault is hit, or you have core dump instrumentation, you can reap the
following benefits:

- **Quicker bug fixes** - If the system is halted or a core dump is captured
  when an assert is hit, the developer can get access to backtraces, registers,
  function arguments, and system state such as the heap, lists, and queues. With
  all this information, developers can root cause issues more quickly.
- **Faster development** - Writing a new piece of code that integrates into
  multiple modules can be tricky if there are bugs or assumptions in the modules
  you are integrating with. By having assertions and OffP practices in the
  surrounding layers, if a developer does anything misaligned with the API
  specs, the system will immediately alert the developer. It's better than
  tracking down what `-3` or `known_error` means and how it was bubbled up
  through the system.

> - **Awareness of prevalence** - No developer or customer notices bugs unless the

Author: Done!

  bug is obvious. By forcing the software to reset, awareness of these undefined
  bugs is increased and developers can now understand how often they are
  occurring.

## Offensive Programming in Practice

Let's dig into some concrete examples about bugs we can find by using Offensive
Programming practices.

### Argument Validation

If a developer comes along and tries to use an API and passes invalid arguments
into the function, make sure the application yells at them to fix the issue.
There is nothing worse than receiving a `-1` return value and digging through 10
layers of firmware code only to find out the real reason was that you passed in
a string of 20 characters when the maximum was 16 or because you accidentally
passed a null pointer due to an uninitialized variable.

```c
void device_set_name(char *name, size_t name_len) {

>   ASSERT(name && name_len =< 16);

Reviewer: ```suggestion
  ASSERT(name && name_len <= 16);
```

  ...
}
```

The only time I would choose, or at least heavily debate, to not use asserts to
validate arguments is when I am building a library that will be used by people
outside of my organization. In those cases, I would make the validation asserts
optional, just as FreeRTOS does with their RTOS functions by allowing developers

> to define ``configASSERT` themselves[^freertos_asserts].

Reviewer: ```suggestion
to define `configASSERT` themselves[^freertos_asserts].
```

### Resource Depletion

Although using dynamic memory in embedded systems is sometimes frowned upon, it
is often necessary to use it on complex systems that don't have enough static
memory to go around. Even when dynamic memory is used, running out of memory
should rarely occur. When memory is low, systems should adapt and put back
pressure on any data flowing into the system and rate-limit memory-intensive
operations.

However, if a firmware _does_ deplete the dynamic memory pool, we want to know
when and why it happened! It could be a memory leak or an accidental allocation
that consumed most of the heap. Running out of memory might not be a
show-stopping issue because the system _might_ recover, but if we are in the
development or internal testing phase, we need to investigate this issue.

To do this, we can add an assert inside of a `malloc()` function that verifies
that the call did not fail.

```c
void *malloc_assert(size_t n) {
  void *p = malloc(n)
  ASSERT(p);
  return p;
  ...
}
```

> In code where the calls to `malloc()` should never fail, such as RTOS structure

Reviewer: ```suggestion
In code where the calls to `malloc()` should never fail, such as for the allocation of RTOS primitives
```

allocation, request and response buffers, etc., we can use this asserted
version.

Another common issue with RTOS-based systems is a queue becoming full and
possibly not emptying quickly enough. As with memory depletion, this isn't a

> show-stopping issue, because the system might recover, but we want to look into

Reviewer: Could call out this means we are dropping events which could be important/critical to operation

Author: Great point

dump the contents of the queue. Then, when the system is halted or you have a
core dump allowing you to find out what events were consuming the majority of
the space in the queue!

```
(gdb) queue_print s_event_queue
Queue Status: 10/10 events in queue (FULL!)
0: Addr: 0x200070c0, event: BLE_PACKET
1: Addr: 0x200070a8, event: TICK_EVENT
2: Addr: 0x20007088, event: BLE_PACKET
3: Addr: 0x20007070, event: BLE_PACKET
4: Addr: 0x20007050, event: BLE_PACKET
5: Addr: 0x20007038, event: BLE_PACKET
6: Addr: 0x20007018, event: BLE_PACKET
7: Addr: 0x20007000, event: BLE_PACKET
8: Addr: 0x20006fe0, event: BLE_PACKET
9: Addr: 0x20006fc8, event: BLE_PACKET
```

Looks like our queue was full of packets from our comms stack and we weren't
processing them quickly enough. We now know where the issue is and we can find a
solution.

### Software Stalls & Deadlocks

Embedded devices need to respond in a reasonable amount of time to user input
and communication packets all while being performant and not showing any signs
of lag or stalls.

I can't count the number of times while working on previous projects that slow
flash operations caused the system to freeze for 2-3 seconds at a time, wreaking

> havoc on the user experience causing other operations in the system to time out.

Reviewer: ```suggestion
havoc on the user experience or causing other operations in the system to time out.
```

The worst part about these issues is that they often aren't brought to our
attention until the end-users complain.

To help catch these issues before pushing firmware to external users, you can

> create and configure your [task

Reviewer: doesn't render as a link

watchdogs]({% post_url 2020-02-18-firmware-watchdog-best-practices %})#adding-a-task-watchdog)
to be more aggressive, set up timers to assert after a few seconds during
potentially long operations, and make sure to set timeouts on your threading
system calls.

```c
void timing_sensitive_task(void) {
  const bool success = mutex_lock(&s_mutex, 1000 /* wait 1s */);
  ASSERT(success);
  {
    ...
  }

> }

Reviewer: if we have a task watchdog, do we need the explicit timeout?

Author: Good point. Added something.

```

When I try to compile this, I get a compile time error assert:

```
$ gcc test.c
test.c:14:1: error: static_assert failed due to requirement
    'sizeof(MyStruct) <= 16' "Oops, too large!"

_Static_assert(sizeof(MyStruct) <= 16, "Oops, too large!");
^              ~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

You can use this on anything that can be statically defined, including alignment
of struct fields, struct and enum sizes, and string lengths. Funny enough, I
also use this when I want to force myself and others to think twice about
certain changes. If I wanted to make sure that I or anyone else

```
_Static_assert(sizeof(MyStruct) == 16,
    "You are changing the size! "
    "Please refer to https://wiki.mycompany.com/... "
    "to learn how to update the protocol version ...");
```

Thankfully, those strings don't get included in the final firmware binary so you
are free to make them as long as you like!

## Handling External & Application Code

There are places where one would not want to use offensive programming
practices. Primarily, it's when the developer is not fully in control of the
software, hardware, or incoming data. These could be any of the following:

> - Hardware & hardware drivers

Reviewer: ```suggestion
- Hardware & peripheral drivers
```

than when a development board is connected to a debugger! With this in place,
developers can instantly and easily get backtraces and system state, as well as
trace through the execution of the undefined behavior to find out what happened
as well as potentially discover what _would_ happen if the bug is not fixed.

### Playing Offense in Production Builds

Before we dig into what aspects of OffP make sense on production builds, we need
to establish that software crashes should be recorded in some way so that a
developer can retrieve them at a later date. This can be some form of logging,
tracing, or core dumps, or any combination of these. If we don't have a way to
record issues, we'll never know what truly causes these issues and how often
they occur.

The parts of OffP that should be kept compiled in and enabled in production
builds are the ones that ensure that undefined behavior is not experienced in
firmware. This includes argument validation, watchdog timeouts to detect
deadlocks, bugs potentially leading to memory corruption, etc. It's better to
return the device to a known state than to limp along.

> What we can turn down to level 3/10 or disable completely are the UX-like bugs,

Reviewer: ```suggestion
What we can often consider disabling completely are marginal issues, such as software stalls, 
```

such as software stalls, malloc and queue failures (which are handled), and
timeouts. These types of issues aren't guaranteed to recover or go away, but
they may be acceptable depending on the context and how the device is used.

If you choose to reduce the aggressiveness of the checks, consider keeping the
hooks and timeouts and instead log them somewhere, rather than reset the system.
For inspiration, check out
[ARM MBed's Error Handling API](https://os.mbed.com/docs/mbed-os/v6.5/apis/error-handling.html),

> which keeps a circular buffer of errors that the system has experienced in a

Reviewer: ```suggestion
which keeps a circular buffer of warnings that the system has experienced in a
```

---
title: Defensive Programming - Friend or Foe?
description:
  Embedded systems can benefit from defensive and offensive programming with the
  use of asserts, timeouts, and watchdogs.
author: tyler
image: img/offensive-programming/cover.png
---

> One of the worst things about embedded development (especially in C) is

Reviewer 2: FWIW I think the slightly shorter introduction you had before was better (but not a big deal).

developers can instantly and easily get backtraces and system state, as well as
trace through the execution of the undefined behavior to find out what happened
as well as potentially discover what _would_ happen if the bug is not fixed.

### Playing Offense in Production Builds

Before we dig into what aspects of OffP make sense on production builds, we need
to establish that software crashes should be recorded in some way so that a
developer can retrieve them at a later date. This can be some form of logging,
tracing, or core dumps, or any combination of these. If we don't have a way to
record issues, we'll never know what truly causes these issues and how often
they occur.

The parts of OffP that should be kept compiled in and enabled in production
builds are the ones that ensure that undefined behavior is not experienced in
firmware. This includes argument validation, watchdog timeouts to detect
deadlocks, bugs potentially leading to memory corruption, etc. It's better to
return the device to a known state than to limp along.

What we can often consider disabling completely are marginal issues, such as

> software stalls, such as software stalls, malloc and queue failures (which are

Reviewer 2: ```suggestion
software stalls, malloc and queue failures (which are
```
